### PR TITLE
c-ares DnsResolverImpl implementation (issue #143).

### PR DESCRIPTION
**Pull request by htuch**
This PR adds c-ares support to DnsResolverImpl, breaking the dependency
on glibc getaddrinfo_a(). It also modifies the DnsResolver API to
support queries that resolve immediately without any deferral and
introduces some more extensive asynch DNS resolution testing via a
simple loopback test DNS server.

### CODE REVIEW

File: source/exe/main.cc [hunk 1]
```diff
@@ -28,6 +30,7 @@ class ProdComponentFactory : public ComponentFactory {
 } // Server
 
 int main(int argc, char** argv) {
+  ares_library_init(ARES_LIB_INIT_ALL);
```
Comment (Reflejo) [inline source/exe/main.cc]: I think ideally this file should be agnostic as to what the underlying implementation of DNS is. FWIW `ares_library_init` is a nop except on windows using winsock. 

It also has some sort of reference counter so you could call `ares_library_init` on the constructor of `DnsResolverImpl` and `ares_library_cleanup` on the destructor. That way the cleanup (if at any point there is any for non windows) will actually happen when the last `DnsResolverImpl` is destroyed.

File: ci/build_container/build_container.sh [hunk 2]
```diff
@@ -65,6 +65,16 @@ make install
 cd ..
 rm -fr nghttp2*
 
+# c-ares
+wget https://github.com/c-ares/c-ares/archive/cares-1_12_0.tar.gz
```
Comment (Reflejo) [inline ci/build_container/build_container.sh]: If you use `https://c-ares.haxx.se/download/c-ares-1.12.0.tar.gz` instead you don't need libtool / automake nor `./buildconf`

File: source/exe/main.cc (hunk 1 above)
Comment (htuch) [inline source/exe/main.cc]: Good point, fixed.

File: ci/build_container/build_container.sh (hunk 2 above)
Comment (htuch) [inline ci/build_container/build_container.sh]: Ah, yeah. I can go either way with this, I originally had a slight preference for using the tar ball connected with the public source and was trying to follow the other source fetches from github. Let me know which one is preferred.

Comment (Reflejo) [inline ci/build_container/build_container.sh]: No strong preference; there are other github tars here so 👌  either way

**Comment by htuch** [comment]
I think Travis is stuck, could someone unstick it? Thanks.

**Comment by Reflejo** [comment]
@htuch trying the good ol' close-reopen trick 🔨

**Comment by mattklein123** [comment]
@htuch I haven't looked at this yet. Is this ready for review? If so can you ping me when it's passing CI.

**Comment by mattklein123** [comment]
@htuch @Reflejo mentions this is not passing because of no c-ares in the image. Can you split adding c-ares to the image into a new PR. We will get that in, then push a new image for you so that this will start passing.

**Comment by htuch** [comment]
@mattklein123 Tests now pass and ready for review. @Reflejo has taken a look and is good with the PR in general.

File: source/common/network/dns_impl.h [hunk 3]
```diff
@@ -5,43 +5,60 @@
...
+  void onAresSocketStateChange(int fd, int read, int write);
+  // Initialize the channel with given ares_init_options().
+  void initializeChannel(struct ares_options* options, int optmask);
 
   Event::DispatcherImpl& dispatcher_;
-  int signal_fd_;
-  event signal_read_event_;
-  std::list<PendingResolutionPtr> pending_resolutions_;
+  ares_channel channel_;
+  std::unordered_map<int, event*> events_;
```
Comment (mattklein123) [inline source/common/network/dns_impl.h]: Please use dispatcher_.createFileEvent(). It will remove a lot of boilerplate and make this code not actually depend on libevent directly. Note that right now FileEvent does not allow setting a timeout. I'm not sure if it's better to modify the the FileEvent interface to add timeout capability, or to just add a parallel Event::Timer for timeouts.

File: source/common/network/dns_impl.cc [hunk 4]
```diff
@@ -5,93 +5,127 @@
...
-                 static_cast<DnsResolverImpl*>(arg)->onSignal();
-               }, this);
-
-  event_add(&signal_read_event_, nullptr);
+  // This is also done in main(), to satisfy the requirement that c-ares is
+  // initialized prior to threading. The additional call to ares_library_init()
+  // here is a nop in normal execution, but exists for testing where we don't
+  // launch via main().
+  ares_library_init(ARES_LIB_INIT_ALL);
+  struct ares_options options;
```
Comment (mattklein123) [inline source/common/network/dns_impl.cc]: nit: del struct (not needed C++)

Comment (mattklein123) [inline source/common/network/dns_impl.cc]: There are a lot of cases of this in the change, if you could remove them all that would be good.

File: test/common/network/dns_impl_test.cc [hunk 5]
```diff
@@ -2,9 +2,232 @@
...
+namespace {
+
+// List of IP address (in human readable format).
+typedef std::list<std::string> IpList;
+// Map from hostname to IpList.
+typedef std::unordered_map<std::string, IpList> HostMap;
+
+// Represents a single TestDnsServer query state and lifecycle. This implements
+// just enough of RFC 1035 to handle queries we generate in the tests below.
+class TestDnsServerQuery {
```
Comment (mattklein123) [inline test/common/network/dns_impl_test.cc]: In general, we are trying to break our dependency on libevent so that eventually it's easier to either use libev, write our own event loop, etc. if we want. We no longer use bufferevent in the prod code. I apologize given how much work you put into this, but please redo this code so that it uses the Envoy interfaces all rooted off of Event::Dispatcher (create listener, create client connections, etc.). You can look at ListenerImplTest to see an example of a a small test that uses real connections. It should be relatively easy to refactor this code around that. Like the FileEvent comment above, it will also get rid of a lot of boiler plate C stuff.

Comment (htuch) [inline test/common/network/dns_impl_test.cc]: Done.

File: source/common/network/dns_impl.cc [hunk 6]
```diff
@@ -1,97 +1,126 @@
 #include "dns_impl.h"
 
 #include "common/common/assert.h"
-#include "common/event/libevent.h"
 #include "common/network/address_impl.h"
 #include "common/network/utility.h"
 
+#include "ares.h"
 #include "event2/event.h"
```
Comment (mattklein123) [inline source/common/network/dns_impl.cc]: should be not needed at this point

File: source/common/network/dns_impl.cc [hunk 7]
```diff
@@ -1,97 +1,126 @@
...
-  signal_fd_ = signalfd(-1, &mask, SFD_NONBLOCK);
-  RELEASE_ASSERT(-1 != signal_fd_);
-
-  event_assign(&signal_read_event_, &dispatcher_.base(), signal_fd_,
-               EV_READ | EV_PERSIST, [](evutil_socket_t, short, void* arg) -> void {
-                 static_cast<DnsResolverImpl*>(arg)->onSignal();
-               }, this);
-
-  event_add(&signal_read_event_, nullptr);
+DnsResolverImpl::DnsResolverImpl(Event::DispatcherImpl& dispatcher)
```
Comment (mattklein123) [inline source/common/network/dns_impl.cc]: Can we make this take Event::Dispatcher vs. Event::DispatcherImpl. Should no longer be dependent on libevent. Also please remove _impl includes for the dispatcher.

File: source/common/network/dns_impl.cc [hunk 8]
```diff
@@ -1,97 +1,126 @@
...
+    delete this;
+    return;
+  }
+  std::list<Address::InstancePtr> address_list;
+  completed_ = true;
+  if (status == ARES_SUCCESS) {
+    ASSERT(hostent->h_addrtype == AF_INET);
+    for (int i = 0; hostent->h_addr_list[i] != nullptr; ++i) {
+      ASSERT(hostent->h_length == sizeof(in_addr));
+      sockaddr_in address;
```
Comment (mattklein123) [inline source/common/network/dns_impl.cc]: just for sanity probably worth memsetting this to zero

File: source/common/network/dns_impl.h [hunk 9]
```diff
@@ -1,47 +1,68 @@
 #pragma once
 
+#include "envoy/event/file_event.h"
 #include "envoy/network/dns.h"
 
 #include "common/common/linked_object.h"
 #include "common/event/dispatcher_impl.h"
```
Comment (mattklein123) [inline source/common/network/dns_impl.h]: per previous comment del

File: source/common/network/dns_impl.h [hunk 10]
```diff
@@ -1,47 +1,68 @@
...
 
+#include "ares.h"
 #include "event2/event_struct.h"
```
Comment (mattklein123) [inline source/common/network/dns_impl.h]: del

File: source/common/network/dns_impl.h [hunk 11]
```diff
@@ -1,47 +1,68 @@
...
 
 namespace Network {
 
+class DnsResolverImplPeer;
```
Comment (mattklein123) [inline source/common/network/dns_impl.h]: Generally prefer not giving test code direct access. Why is this necessary? Can we refactor to make it not needed?

File: source/common/network/dns_impl.cc [hunk 12]
```diff
@@ -1,97 +1,126 @@
...
+  };
+  options->sock_state_cb_data = this;
+  ares_init_options(&channel_, options, optmask | ARES_OPT_SOCK_STATE_CB);
+}
 
-    std::list<Address::InstancePtr> address_list;
-    addrinfo* result = pending_resolution->async_cb_data_.ar_result;
-    while (result != nullptr) {
+void DnsResolverImpl::PendingResolution::onAresHostCallback(int status, hostent* hostent) {
+  if (status == ARES_EDESTRUCTION) {
```
Comment (mattklein123) [inline source/common/network/dns_impl.cc]: This owned business is a little confusing. I'm guessing ARES_EDESTRUCTION happens when we destroy the channel with a pending request? If so, can we just put a comment to that effect?

File: source/common/network/dns_impl.h (hunk 11 above)
Comment (mattklein123) [inline source/common/network/dns_impl.h]: I think I see that you roughly need to tweak the resolver behavior for the tests. Can we make the constructor take params? Or have some virtual function which a test resolver overrides?

Comment (htuch) [inline source/common/network/dns_impl.h]: At Google, we tend to use this FooPeer pattern for reaching into the implementation internals. I'll look at refactoring, but it seems a useful tool to have available when writing tests in general, avoiding introducing test specific behavior into the implementation.

Comment (mattklein123) [inline source/common/network/dns_impl.h]: If it's not easy to refactor or would make the code substantially worse, then I would skip it. It's just not a pattern that we are currently using anywhere. If it's straightforward it would be good to do for consistency reasons.

Comment (htuch) [inline source/common/network/dns_impl.h]: I pushed this into a resetChannelForTest() method. Most of this is really very test specific, I don't know if it makes that much sense to put behavior like specifying the server:port CSV into a non-test method. I'm good either way, I can remove the final patch if you prefer.

Comment (mattklein123) [inline source/common/network/dns_impl.h]: nah, I think if we are doing something like this I agree the peer approach is better. sorry for the runaround, let's just revert his last patch.

Comment (htuch) [inline source/common/network/dns_impl.h]: Alright. I'll keep in mind the preference to avoid peers in future.
